### ISSUE TITLE
Tag management is merging some words

### ISSUE THREAD

**Issue by EtienneAb3d** [modernmt/modernmt#510]
I think this problem was already there with older MMT 3.

When I'm translating this sentence (MS Word XML with tags renamed by my process):
`MMT befxml : <LbxTag001>Par<LbxTag002>nature ,<LbxTag003>l' atteinte<LbxTag004>de<LbxTag005>ces<LbxTag006>objectifs est<LbxTag007>soumise<LbxTag008>à<LbxTag009>de<LbxTag010>nombreux<LbxTag011>risques<LbxTag012>et<LbxTag013>incertitudes<LbxTag014>susceptibles.<LbxTag015>`

The translation shown in the log file is correct:
`   translation = By nature , the achievement of these objectives is subject to many risks and uncertainties that may arise .`

But, the final translation with tags is merging 'theseobjectives':
`{"status":200,"data":{"translation":"<LbxTag001>By<LbxTag002>nature<LbxTag003>, the<LbxTag004><LbxTag005><LbxTag006> achievement of theseobjectives<LbxTag007><LbxTag008> issubject to<LbxTag009><LbxTag010> many<LbxTag011>risks<LbxTag012>and<LbxTag013>uncertainties<LbxTag014>that may arise.<LbxTag015>","sourceWordCount":21,"targetWordCount":19,"contextVector":{"entries":[{"memory":{"id":1,"name":"USERS.fr-en"},"score":0.09509246}]},"priority":"normal","totalTime":316,"memoryLookupTime":2,"decodingTime":308,"queueTime":0,"queueLength":0}}`

PS: updated to show exactly what was sent to MMT

**Comment by EtienneAb3d** [comment]
Forget this: I see that the merging was already in the original FR sentence...

**Comment by EtienneAb3d** [comment]
After fixing my own code, the problem is still there.

Sent to MMT4:
`<LbxTag001>Par<LbxTag002> nature ,<LbxTag003> l' atteinte<LbxTag004> de<LbxTag005> ces<LbxTag006> objectifs est<LbxTag007> soumise<LbxTag008> à<LbxTag009> de<LbxTag010> nombreux<LbxTag011> risques<LbxTag012> et<LbxTag013> incertitudes<LbxTag014> susceptibles.<LbxTag015>`

MMT4 LOG (correct):
`   translation = By nature , the achievement of these objectives is subject to many risks and uncertainties that may arise .`
 
Final MMT4 reply(word mergin on 'theseobjectives'): 
`{"status":200,"data":{"translation":"<LbxTag001>By<LbxTag002> nature<LbxTag003> , the<LbxTag004><LbxTag005><LbxTag006> achievement of theseobjectives<LbxTag007><LbxTag008> issubject to<LbxTag009><LbxTag010> many<LbxTag011> risks<LbxTag012> and<LbxTag013> uncertainties<LbxTag014> that may arise.<LbxTag015>","sourceWordCount":21,"targetWordCount":19,"contextVector":{"entries":[{"memory":{"id":1,"name":"USERS.fr-en"},"score":0.09509246}]},"priority":"normal","totalTime":546,"memoryLookupTime":22,"decodingTime":336,"queueTime":0,"queueLength":0}}`

**Comment by nicolabertoldi** [comment]
@EtienneAb3d 

we will investigate soon.

in the node log (if you start the engine with the parameter "-v 2",
you should find some additional info related to the actual translation computed by ModernMT
similar to
```
sentence = Payment
   translation = Pagamento
   alignment = 0-0
   suggestions = [
]
```

For the "problematic" sentence, could you please report such info.

It is correct that tags are not reported there.

**Comment by EtienneAb3d** [comment]
Here it is (I added a code at the beginning of the source text, see #452).
The bad sentence found in the memory is due to the history of translation management, using a previous bad translation made by MMT. But, as you can see, this bad memory entry doesn't produce the bad translation result. The translation result before tag processing is correct.
```
   sentence = # A # Par nature , l' atteinte de ces objectifs est soumise à de nombreux risques et incertitudes susceptibles .
   translation = By nature , the achievement of these objectives is subject to many risks and uncertainties that may arise .
   alignment = 0-10 0-14 0-18 1-10 1-14 1-18 2-10 2-14 2-18 3-0 4-1 6-4 7-2 7-3 7-4 9-6 10-4 10-7 11-8 12-8 12-9 13-8 15-11 16-12 17-13 18-14 19-15 20-17
   suggestions = [
      ScoreEntry{memory=1, language=fr \u2192 en, sentence=[#, G, #, Par, nature, ,, l', atteinte, de, ces, objectifs, est, soumise, à, de, nombreux, risques, et, incertitudes, susceptibles], translation=[By, nature, ,, the, achievement, of, theseobjectives, issubject, to, many, risks, and, uncertainties, that, may, arise], score=0.48773745}
   ]

```

**Comment by EtienneAb3d** [comment]
As said, I think the problem was already in MMT 3.
Here is a comparison: left column was translated with MMT 3, while right column was translated with MMT 4.
![image](https://user-images.githubusercontent.com/25932245/66129951-2e68a600-e5f1-11e9-9b44-0206e0d94f7e.png)

**Comment by nicolabertoldi** [comment]
@EtienneAb3d 

we will check asap

**Comment by nicolabertoldi** [comment]
@EtienneAb3d 

are you sure that you reported the log snipper corresponding to the problematic sentence?
you reported
```
sentence = # A # Par nature , l' atteinte de ces objectifs est soumise à de nombreux risques et incertitudes susceptibles .
```
but actually in your input 
```
MMT befxml : <LbxTag001>Par<LbxTag002>nature ,<LbxTag003>l' atteinte<LbxTag004>de<LbxTag005>ces<LbxTag006>objectifs est<LbxTag007>soumise<LbxTag008>à<LbxTag009>de<LbxTag010>nombreux<LbxTag011>risques<LbxTag012>et<LbxTag013>incertitudes<LbxTag014>susceptibles.<LbxTag015>
```
the text `# A #` is missing.

Could you please double check?

If you translate the same sentence without Lbxtags, which translation and alignment did you get?

**Comment by EtienneAb3d** [comment]
The "# A #" is added just before sending to MMT. The source sentence I provided without it is just before this add.

**Comment by nicolabertoldi** [comment]
@EtienneAb3d 

I just finished to improve the tag projection.

According to my debugging, all problems you experienced are now solved.
But before pushing the change in the master, I would kindly ask that you double-check as well.

So, could you please pull the branch "features/improving_tag_projection", using this branch test on your data, and report to me.

I am sure you know how to setup and compile a branch different from the master.
Otherwise please contact me privately.

**Comment by nicolabertoldi** [comment]
@EtienneAb3d 
I have not changed the regression unit test yet.

Please, exclude the testing suite from the mvn compilation using the parameter `-Dmaven.test.skip=true`.

Basically, you can run
```
mvn -Dmaven.test.skip=true clean install
```

**Comment by EtienneAb3d** [comment]
OK ! On a test sentence, this seems to solve the problem. 
:)

**Comment by nicolabertoldi** [comment]
@EtienneAb3d 

If you would like to verify that the new version is better on a larger test set, you are welcome.

**Comment by EtienneAb3d** [comment]
Bad news, I got 2 problems:
1) many strange ">" added. Example:

Sent to MMT4:
`http://localhost:10001/translate?q=%23E%23+%3CLbxTag001%3ERecherche%3CLbxTag002%3E+et%3CLbxTag003%3E+innovation+%3CLbxTag004%3E33.%3CLbxTag005%3E&source=fr&target=en&context=assembl%C3%83%C2%A9e+AG+personnel+gestion+comptabilit%C3%83%C2%A9`

MMT4 LOG:
```
2019-10-21 15:19:22,140 DEBUG eu.modernmt.decoder.neural.NeuralDecoder [pool-2-thread-4] Translation received from neural decoder:
   sentence = # E # Recherche et innovation 33 .
   translation = Research and innovation 33 .
   alignment = 0-4 1-4 2-4 3-0 4-1 5-2 6-3 7-4
   suggestions = [
   ]

Result (see "innovation>"):
{"status":200,"data":{"translation":"<LbxTag001>Research<LbxTag002> and<LbxTag003> innovation><LbxTag004>33.<LbxTag005>","sourceWordCount":8,"targetWordCount":5,"contextVector":{"entries":[]},"priority":"normal","totalTime":74,"memoryLookupTime":0,"decodingTime":66,"queueTime":0,"queueLength":0}}

2) still one merged word

Sent to MMT4:
`http://localhost:10001/translate?q=%23C%23+%3CLbxTag001%3Ed%C3%A9mocratique+%3CLbxTag002%3Edu+Congo.%3CLbxTag003%3E&source=fr&target=en&context=assembl%C3%83%C2%A9e+AG+personnel+gestion+comptabilit%C3%83%C2%A9`

MMT4 LOG:
```
   sentence = # C # démocratique du Congo .
   translation = Democratic Republic of the Congo .
   alignment = 0-1 0-2 0-4 1-1 1-2 1-4 2-1 2-2 2-4 3-0 5-3 6-5
   suggestions = [
   ]

Result (see "Republicof"):
{"status":200,"data":{"translation":"<LbxTag001>Democratic Republicof <LbxTag002>the Congo.<LbxTag003>","sourceWordCount":7,"targetWordCount":6,"contextVector":{"entries":[]},"priority":"normal","totalTime":93,"memoryLookupTime":0,"decodingTime":85,"queueTime":0,"queueLength":0}}

```

**Comment by nicolabertoldi** [comment]
@EtienneAb3d 

I will debug soon.

Could you please confirm that the input strings are
 ```
#E# <LbxTag001>Recherche<LbxTag002> et<LbxTag003> innovation <LbxTag004>33.<LbxTag005>
#C# <LbxTag001>démocratique <LbxTag002>du Congo.<LbxTag003>
```

**Comment by EtienneAb3d** [comment]
Yes.
But, I find something more strange with the second example : words are merge in both case, with or without entry tags. I mean, the LOG provides with the good (not merged) translation, with or without entry tags, but words are merged even when no tag is provided in the sentence to be translated. This is observed with both the official MMT4 and the new patch.

**Comment by nicolabertoldi** [comment]
@EtienneAb3d 

the way ModernMT handles and re-insert spaces and tags is strictly related to the alignment between the tokens.
And in  your second case the alignment is not very good.

By the way, having these two more examples, I should be able to debug and solve some special cases.

One more point.
I noticed that in all your examples your input start with a string like `#E#` but they are not produced in the translations. This suggests that in your training data you have such string in the source part, but not in the target.

ModernMT works fine even in this case, but it is not advisable to have such differences in the training data.
So if you add on your own that string, I would suggest to add it on both source and target sides.

**Comment by EtienneAb3d** [comment]
You are right: the training data are all with source marked and target not marked.

**Comment by EtienneAb3d** [comment]
The reason is that, I would like the model to learn this mark as a global influencer of the result. My intuition, is that, marking both sides could bring me with a different learning, where a small part of the network will quickly learn to copy this mark, while the whole network won't take it into account any more since no more included in the error adaptation.

**Comment by nicolabertoldi** [comment]
@EtienneAb3d 

Regarding the first problem (`innovation>`).
I just committed a fix.


Regarding the second problem (`republicof`)
As said, our policy to re-insert spaces and tags is strongly based on the word-alignment between source and target words. In this case the WA is quite bad due to the presence of the "additional" item `#C#` which actually corresponds to three tokens `#`, `C`, and `#`.

If you still want to use a marker I would suggest to use a different one which is not tokenized, like for instance `MARK-C`.

At the moment we cannot modify this behaviour, because it would require a full re-structuring of our post-processing policy.
But surely we will take into account this issue in the future.

**Comment by EtienneAb3d** [comment]
For the second problem, what I do not understand is why you are removing spaces. The translation is good before tag processing. Inserting tag should not change the spaces.

**Comment by nicolabertoldi** [comment]
@EtienneAb3d 

the problem is not related to the tags, but to the alignment.
In fact the same issue arise without tags as you observed.

Practically, the two target tokens `republic` and `of` are aligned (wrongly) with the two source tokens `C` and `#`, respectively.
As we trust on the original spacing between `C` and `#`, which  is not present, we do not put any space between `republic` and `of`.

This policy works almost perfectly for handling the spacing before and after punctuation and tags, and for all other tokens when alignment is good.

**Comment by EtienneAb3d** [comment]
Except for very few word mergings, the new update seems good now. It's a little bit better than what I got with MMT3. But, this is possibly due to my new tuning code addition : there are less lost words, and this is exactly the reason why it is done (see #452).

**Comment by nicolabertoldi** [comment]
@EtienneAb3d 

good to know

If you think it is fine, I would kindly ask you to close this issue.

Please feel free to open a new one if needed in the future.

**Comment by EtienneAb3d** [comment]
You need to commit the new change in the official delivery. Don't forget to tune the test step in the install script.

**Comment by mzeidhassan** [comment]
@nicolabertoldi  I am sorry for posting here again. I am wondering if this is already merged in the official repo?
